[]json.Number decoding broken on arm64

**Describe the bug**

Sonic 1.13.1 fails to decode arrays of numbers to `[]json.Number` on arm64. `encoding/json` handles this.

**To Reproduce**

```go
package sonictest

import (
	"encoding/json"
	"testing"

	"github.com/bytedance/sonic"
)

type numberArray struct {
	X []json.Number `json:"x"`
}

const jsonInput = "{\"x\": [1, 2, 3]}"

func TestDecoding(t *testing.T) {
	var stdlibOutput numberArray
	err := json.Unmarshal([]byte(jsonInput), &stdlibOutput)
	if err != nil {
		t.Error("stdlib json unmarshal error:", err)
	}

	var sonicOutput numberArray
	err = sonic.Unmarshal([]byte(jsonInput), &sonicOutput)
	if err != nil {
		t.Error("sonic json unmarshal error:", err)
	}
}
```
Output:
```
=== RUN   TestDecoding
    sonic_test.go:26: sonic json unmarshal error: Mismatch type string with value number "at index 7: mismatched type with value\n\n\t{\"x\": [1, 2, 3]}\n\t.......^........\n"
--- FAIL: TestDecoding (0.00s)
```

**Expected behavior**
Sonic populates x with a series of json.Number values.

**Sonic version:**

v1.31.1

**Environment:**

```
AR='ar'
CC='cc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='c++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/heen/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/heen/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/0h/jys8cw2d34nfly3ldt2t84x80000gn/T/go-build2502036572=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/Users/heen/dev/frontend/highfive/go.mod'
GOMODCACHE='/Users/heen/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/heen/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.24.0/libexec'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/Users/heen/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.24.0/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.24.0'
GOWORK=''
PKG_CONFIG='pkg-config'
```

--- a/internal/decoder/optdec/compiler.go
+++ b/internal/decoder/optdec/compiler.go
@@ -1,7 +1,6 @@
 package optdec
 
 import (
-	"encoding/json"
 	"fmt"
 	"reflect"
 
@@ -265,7 +264,7 @@ func (c *compiler) compileSlice(vt reflect.Type) decFunc {
 	if et.IsUint64() {
 		return &sliceU64Decoder{}
 	}
-	if et.Kind() == reflect.String {
+	if et.Kind() == reflect.String && et != rt.JsonNumberType {
 		return &sliceStringDecoder{}
 	}
 
@@ -345,7 +344,7 @@ func (c *compiler) compileMap(vt reflect.Type) decFunc {
 	// Some common integer map later
 	mt := rt.MapType(rt.UnpackType(vt))
 
-	if mt.Key.Kind() == reflect.String {
+	if mt.Key.Kind() == reflect.String && mt.Key != rt.JsonNumberType {
 		return &mapStrKeyDecoder{
 			mapType: mt,
 			assign: rt.GetMapStrAssign(vt),
@@ -420,8 +419,13 @@ func (c *compiler) compileMapKey(vt reflect.Type) decKey {
 		return decodeFloat32Key
 	case reflect.Float64:
 		return decodeFloat64Key
+	case reflect.String:
+		if rt.UnpackType(vt.Key()) == rt.JsonNumberType {
+			return decodeJsonNumberKey
+		}
+		fallthrough
 	default:
-		panic(&json.UnmarshalTypeError{Type: vt})
+		return nil
 	}
 }
 

--- a/internal/decoder/optdec/map.go
+++ b/internal/decoder/optdec/map.go
@@ -2,6 +2,7 @@ package optdec
 
 import (
 	"encoding"
+	"encoding/json"
 	"math"
 	"reflect"
 	"unsafe"
@@ -291,9 +292,9 @@ func (d *mapU64KeyDecoder) FromDom(vp unsafe.Pointer, node Node, ctx *context) e
 
 /** Decoder for generic cases */
 
-type decKey func(dec *mapDecoder, raw string, ctx *context) (interface{}, error)
+type decKey func(dec *mapDecoder, raw string) (interface{}, error)
 
-func decodeKeyU8(dec *mapDecoder, raw string, ctx *context) (interface{}, error) {
+func decodeKeyU8(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
@@ -308,7 +309,7 @@ func decodeKeyU8(dec *mapDecoder, raw string, ctx *context) (interface{}, error)
 	return uint8(ret), nil
 }
 
-func decodeKeyU16(dec *mapDecoder, raw string, ctx *context) (interface{}, error) {
+func decodeKeyU16(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
@@ -323,7 +324,7 @@ func decodeKeyU16(dec *mapDecoder, raw string, ctx *context) (interface{}, error
 	return uint16(ret), nil
 }
 
-func decodeKeyI8(dec *mapDecoder, raw string, ctx *context) (interface{}, error) {
+func decodeKeyI8(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
@@ -338,7 +339,7 @@ func decodeKeyI8(dec *mapDecoder, raw string, ctx *context) (interface{}, error)
 	return int8(ret), nil
 }
 
-func decodeKeyI16(dec *mapDecoder, raw string, ctx *context) (interface{}, error) {
+func decodeKeyI16(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
@@ -353,7 +354,7 @@ func decodeKeyI16(dec *mapDecoder, raw string, ctx *context) (interface{}, error
 	return int16(ret), nil
 }
 
-func decodeKeyTextUnmarshaler(dec *mapDecoder, raw string, ctx *context) (interface{}, error) {
+func decodeKeyTextUnmarshaler(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
@@ -366,7 +367,7 @@ func decodeKeyTextUnmarshaler(dec *mapDecoder, raw string, ctx *context) (interf
 	return ret, nil
 }
 
-func decodeFloat32Key(dec *mapDecoder, raw string, _ *context) (interface{}, error) {
+func decodeFloat32Key(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
@@ -381,12 +382,25 @@ func decodeFloat32Key(dec *mapDecoder, raw string, _ *context) (interface{}, err
 	return float32(ret), nil
 }
 
-func decodeFloat64Key(dec *mapDecoder, raw string, _ *context) (interface{}, error) {
+func decodeFloat64Key(dec *mapDecoder, raw string) (interface{}, error) {
 	key, err := Unquote(raw)
 	if err != nil {
 		return nil, err
 	}
 	return ParseF64(key)
+}
+
+func decodeJsonNumberKey(dec *mapDecoder, raw string) (interface{}, error) {
+	// skip the quote
+	raw = raw[1:len(raw)-1]
+	end, ok := SkipNumberFast(raw, 0)
+
+	// check trailing chars
+	if !ok || end != len(raw) {
+		return nil, error_value(raw, rt.JsonNumberType.Pack())
+	}
+	
+	return json.Number(raw[0:end]), nil
 }
 
 type mapDecoder struct {
@@ -402,8 +416,8 @@ func (d *mapDecoder) FromDom(vp unsafe.Pointer, node Node, ctx *context) error {
 	}
 
 	obj, ok := node.AsObj()
-	if !ok {
-		return	error_mismatch(node, ctx, d.mapType.Pack())
+	if !ok || d.keyDec == nil {
+		return error_mismatch(node, ctx, d.mapType.Pack())
 	}
 
 	// allocate map
@@ -417,7 +431,8 @@ func (d *mapDecoder) FromDom(vp unsafe.Pointer, node Node, ctx *context) error {
 	for i := 0; i < obj.Len(); i++ {
 		keyn := NewNode(next)
 		raw := keyn.AsRaw(ctx)
-		key, err := d.keyDec(d, raw, ctx)
+
+		key, err := d.keyDec(d, raw)
 		if err != nil {
 			if gerr == nil {
 				gerr = error_mismatch(keyn, ctx, d.mapType.Pack())

--- a/issue_test/common_test.go
+++ b/issue_test/common_test.go
@@ -36,7 +36,7 @@ func assertUnmarshal(t *testing.T, api sonic.API, cas unmTestCase, args ...inter
 	assert.Equal(t, jv, sv, spew.Sdump(jv, sv))
 	assert.Equal(t, serr == nil, jerr == nil, spew.Sdump(jerr, serr))
 	if len(args) > 0 && args[0].(bool) {
-		spew.Dump(sv, serr, jv, jerr)
+		spew.Dump(jv, jerr, sv, serr)
 	}
 }
 

--- a/issue_test/issue777_test.go
+++ b/issue_test/issue777_test.go
@@ -1,0 +1,76 @@
+/*
+ * Copyright 2022 ByteDance Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ package issue_test
+
+ import (
+	"encoding/json"
+	 "testing"
+ 
+	 "github.com/bytedance/sonic"
+ )
+ 
+ type unknownKeyMap struct {
+	 X []json.Number `json:"x"`
+	 Y map[*int]json.Number `json:"y"`
+	 Z map[int]func() `json:"z"`
+ }
+ 
+ func TestIssue777_NumberSlice(t *testing.T) {
+	cas := unmTestCase {
+		name: "number slice",
+		data: []byte(`["1", "2", 123, 456]`),
+		newfn: func() interface{} {
+			var a []json.Number
+			return &a
+		},
+	}
+	assertUnmarshal(t, sonic.ConfigStd, cas)
+}
+
+ func TestIssue777_NumberKey(t *testing.T) {
+	 cas := unmTestCase {
+		 name: "number key",
+		 data: []byte(`{"1": "2", "123": 123}`),
+		 newfn: func() interface{} {
+			 var a map[json.Number]json.Number
+			 return &a
+		 },
+	 }
+	 assertUnmarshal(t, sonic.ConfigStd, cas)
+ 
+	 // TODO: encoding/json has a bug
+	 // cas = unmTestCase {
+	 // 	name: "number key",
+	 // 	data: []byte(`{"1": "2", "123x": 123}`),
+	 // 	newfn: func() interface{} {
+	 // 		var a map[json.Number]json.Number
+	 // 		return &a
+	 // 	},
+	 // }
+	 // assertUnmarshal(t, sonic.ConfigStd, cas, true)
+ 
+	// TODO: jit has a bug here
+	//  cas = unmTestCase {
+	// 	 name: "skip unknown key map",
+	// 	 data: []byte("{\"z\":{}, \"y\": {\"1\": \"2\", \"123\": 123}, \"x\": [1, 2, 3]}"),
+	// 	 newfn: func() interface{} {
+	// 		 var a unknownKeyMap
+	// 		 return &a
+	// 	 },
+	//  }
+	//  assertUnmarshal(t, sonic.ConfigStd, cas)
+ }

